To convert the clonal phylogeny from a relative tree representing genetic differences into a time-scaled chronogram, we conducted a Bayesian molecular clock analysis. The primary goal of this analysis is to estimate the specific divergence times for the metastatic clonal lineages within the tumor.


## Step 1: Data Preparation 

clonal sequences alignment were generated by CloneFinder

## Step 2: Bayesian Phylogenetic Inference with BEAST

### 2.1 nucleotide substitution model selection with ModelFinder

```sh
iqtree2 -s res/wgs_clonal_molecular_dating/case1_clonefinder.fa \
	 -m MFP -mset beast2 \
	 -o NORMAL   -B 1000 \
	 --prefix  res/wgs_clonal_molecular_dating/case1_model_test

```

Best-fit model according to BIC for both cases: GTR+F+ASC+R2


To fit the "+ASC", we modified Site Model in xml file for case1

```xml
<alignment id="case1" spec="FilteredAlignment" filter="-">
							
	<data idref="case1Original"/>
							
	<constantSiteWeights id="IntegerParameter.0" spec="parameter.IntegerParameter" dimension="4" lower="0" upper="0">25831188 17876406 17876406 25831188</constantSiteWeights>
						
</alignment>
```


and modified xml file for case2
```xml
<alignment id="case2" spec="FilteredAlignment" filter="-">
							
	<data idref="case2Original"/>
							
	<constantSiteWeights id="IntegerParameter.0" spec="parameter.IntegerParameter" dimension="4" lower="0" upper="0">18502572 12804657 12804657 18502572</constantSiteWeights>
						
</alignment>
```


### 2.2 Clock Model & Tree Prior selection using Path Sampling

using the Path Sampling method, we compared several candidate models:

- Clock Model: strict clock, optimised relaxed clock, random local clock, relaxed exponential, and relaxed log normal 
    
- Tree Prior: constant population size, Bayesian skyline, and exponential growth 

a total of 15 models combinations were test for each case, Four XML files were created using BEAUti (v2.6.7) , load sequence alignment and  set to Site Model to "GTR+F+ASC+R2" , and set corresponding Tree Prior and Clock Model. 

each model combinations were applied in BEAST (v2.6.7) run MCMC with 100 steps and each with 1000000 chain length. The Marginal_Likelihood_Value were calculated and recorded for each model combination. 



| Case  | Tree Prior  | Clock Model      | Log Marginal Likelihood | 2 * Diff ML |
| ----- | ----------- | ---------------- | ----------------------- | ----------- |
| Case1 | exponential | optClock         | -119964639.4            | 10941.23    |
| Case1 | constant    | optClock         | -119970110.1            | 112.36      |
| Case1 | exponential | logClock         | -119970166.2            | 83.18       |
| Case1 | bayes       | strictClock      | -119970207.8            | 0.46        |
| Case1 | bayes       | randomClock      | -119970208.1            | 2.6         |
| Case1 | exponential | randomClock      | -119970209.4            | 0.36        |
| Case1 | exponential | strictClock      | -119970209.5            | 1.03        |
| Case1 | constant    | randomClock      | -119970210.1            | 0.07        |
| Case1 | constant    | strictClock      | -119970210.1            | 1.54        |
| Case1 | bayes       | optClock         | -119970210.9            | 435.54      |
| Case1 | constant    | logClock         | -119970428.6            | 1434.97     |
| Case1 | bayes       | logClock         | -119971146.1            | 6707.73     |
| Case1 | bayes       | exponentialClock | -119974500              | 1856        |
| Case1 | exponential | exponentialClock | -119975428              | 4321.46     |
| Case1 | constant    | exponentialClock | -119977588.7            |             |

| Case  | Tree Prior  | Clock Model      | Log Marginal Likelihood | 2 * Diff ML |
| ----- | ----------- | ---------------- | ----------------------- | ----------- |
| Case2 | exponential | optClock         | -139796175.9            | 851.67      |
| Case2 | bayes       | logClock         | -139796601.7            | 14964.52    |
| Case2 | constant    | randomClock      | -139804084              | 1342.27     |
| Case2 | constant    | logClock         | -139804755.1            | 423.84      |
| Case2 | exponential | logClock         | -139804967              | 224.17      |
| Case2 | bayes       | randomClock      | -139805079.1            | 0.1         |
| Case2 | bayes       | strictClock      | -139805079.1            | 8.8         |
| Case2 | bayes       | optClock         | -139805083.5            | 0.83        |
| Case2 | constant    | strictClock      | -139805084              | 1.72        |
| Case2 | exponential | strictClock      | -139805084.8            | 0.1         |
| Case2 | exponential | randomClock      | -139805084.9            | 5.99        |
| Case2 | constant    | optClock         | -139805087.9            | 3259.5      |
| Case2 | constant    | exponentialClock | -139806717.6            | 15049.06    |
| Case2 | exponential | exponentialClock | -139814242.1            | 8893.77     |
| Case2 | bayes       | exponentialClock | -139818689              |             |


A Model Combination (exponential + optClock)  is favor for both cases with strong evidence (2 * Diff ML >10) 


### 2.3 MCMC analysis and Post-processing

After selecting the best-fit models, the final, full-length MCMC (Markov Chain Monte Carlo) analysis is performed to generate the posterior distribution of trees and parameters. This is the main computational step that produces the results for downstream analysis.

The analysis was run in **BEAST (v2.6.7)** using the optimal model combination determined through path sampling:

- **Site Model**: **GTR+F+ASC+R2**
    
- **Clock Model: optimised relaxed clock**, with a middle ORCulcdMean of 2.3e-9 substitutions per site per generation
    
- **Tree Prior**: **Exponential Growth Population Size**
    
- **MRCA Prior**:All tumor clones were constrained as a monophyletic group, and a uniform prior was placed on the age of their most recent common ancestor (MRCA) for two separate cases: 0–12,775 generations (Case 1) and 0–6,388 generations (Case 2), assuming a mean generation time of two days.


The MCMC chain was run for 5e8 generations, with both parameters and trees logged every 2000 generations.  


Convergence of the MCMC analysis was assessed in Tracer (v1.7.1) to ensure the reliability of the results with the Effective Sample Size (ESS) above 200. The MCC tree was calculated using TreeAnnotator with 20% trees Burn-in and a median node height.



## Step 3: Quantification of Relative Metastasis Timing

To determine if metastasis was an early or late event in the tumor's evolution, we calculated a normalized timing metric for each tree in the posterior distribution from our BEAST analysis.

$$\text{Relative Timing} = \frac{t_{MRCA} - m_{MRCA}}{t_{MRCA}}$$

Where:
*   **$t_{MRCA}$**: The time to the Most Recent Common Ancestor of **all** tumor cells (both primary and metastatic). This value represents the total evolutionary lifespan of the tumor, from its origin to the present.
*   **$m_{MRCA}$**: The time to the Most Recent Common Ancestor of the **metastatic** cells only. This value marks the evolutionary origin of the metastatic lineage.


The relative timing metric was computed for each tree within the posterior distribution generated by the **BEAST** software. An initial burn-in period from the Markov chain Monte Carlo (MCMC) run was discarded to ensure that calculations were performed only on stable posterior samples. Finally, by applying this formula across all trees in the posterior distribution, a full posterior probability distribution for the relative timing metric is generated, effectively capturing the statistical uncertainty in the estimate.
 